File: leetcode/find-the-difference-of-two-arrays.py
# 2215. Find the Difference of Two Arrays
# 🟢 Easy
#
# https://leetcode.com/problems/find-the-difference-of-two-arrays/
#
# Tags: Array - Hash Table

import timeit
from typing import List


# What they are really asking for is a tuple with set set differences of
# a-b and b-a, but in a list format, we can return just that.
#
# Time complexity: O(n) - We iterate all the elements multiple times.
# Space complexity: O(n) - We are creating sets of the same size as the
# input.
#
# Runtime 182 ms Beats 61.53%
# Memory 16.8 MB Beats 9.53%
class Solution:
    def findDifference(
        self, nums1: List[int], nums2: List[int]
    ) -> List[List[int]]:
        a, b = set(nums1), set(nums2)
        return [list(a - b), list(b - a)]


def test():
    executors = [Solution]
    tests = [
        [[1, 2, 3, 3], [1, 1, 2, 2], [[3], []]],
        [[1, 2, 3], [2, 4, 6], [[1, 3], [4, 6]]],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.findDifference(t[0], t[1])
                exp = t[2]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/find-the-difference-of-two-arrays.rs
// 2215. Find the Difference of Two Arrays
// 🟢 Easy
//
// https://leetcode.com/problems/find-the-difference-of-two-arrays/
//
// Tags: Array - Hash Table

use std::collections::HashSet;

struct Solution;
impl Solution {
    /// Get two sets with the elements of nums1 and nums2 respectively, then
    /// get the differences between sets as vectors.
    ///
    /// Time complexity: O(n) - We iterate over all values and do O(1) work.
    /// Space complexity: O(n) - We use two hash sets of size n of extra memory.
    ///
    /// Runtime 14 ms Beats 47.62%
    /// Memory 2.3 MB Beats 42.86%
    pub fn find_difference(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<Vec<i32>> {
        let sa = nums1.iter().cloned().collect::<HashSet<_>>();
        let sb = nums2.iter().cloned().collect::<HashSet<_>>();
        vec![
            (&sa - &sb).iter().cloned().collect::<Vec<_>>(),
            (&sb - &sa).iter().cloned().collect::<Vec<_>>(),
        ]
    }
}

// Tests.
fn main() {
    let tests = [
        (vec![1, 2, 3, 3], vec![1, 1, 2, 2], vec![vec![3], vec![]]),
        (vec![1, 2, 3], vec![2, 4, 6], vec![vec![3, 1], vec![6, 4]]),
    ];
    for t in tests {
        assert_eq!(Solution::find_difference(t.0, t.1), t.2);
    }
    println!("\x1b[92m» All tests passed!\x1b[0m")
}
